odoo/tools/translate.py

```diff
@@ -259,18 +264,25 @@ def process(node):
             result.tail = node.tail
             has_text = (
                 todo_has_text or nonspace(result.text) or nonspace(result.tail)
-                or any((key in TRANSLATED_ATTRS and val) for key, val in result.attrib.items())
+                or any((key in TRANSLATED_ATTRS and val) for key, val in attrib)
             )
             return (has_text, result)
 
         # translate the content of todo and append it to result
         append_content(result, translate_content(todo) if todo_has_text else todo)
 
         # translate the required attributes
-        for name, value in result.attrib.items():
+        for name, value in attrib:
             if name in TRANSLATED_ATTRS:
                 result.set(name, translate_text(value) or value)
 
+        # translate the attributes on some unique border cases.
+        for xpath, attribute in TRANSLATED_XPATH_ATTRS:
+            if result.xpath(xpath):
```

Comment:
That doesn't seem correct: because the xpath expression is absolute, it will check if the node is part of a document which matches the path, not anything about the node itself.

I'm also not quite sure what it's trying to do, is it trying to translate the `@value` but only if it's an input of type text without the `datetimepicker-input` class?

Reply:
That also duplicates the TRANSLATED_ATTRS loop whereas it seems to me we could just... update the protocol of TRANSLATED_ATTRS to add an optional predicate?

Reply:
And thinking about the issue further, there seem to be two more issues with it:
* `text` is the default value for `input/@type`, so should `input` elements *without* a `@type` also be selected?

And `textarea` are the multi-line dual to text inputs, should we not also translate that then?

Reply:
@xmo-odoo 

> That doesn't seem correct: because the xpath expression is absolute [...]

Not an expert here but since `result` is a unique node, I think it is ok? I tested (again and again) and it still seems to work. If you prefer, instead of xpath expressions, I could use lambda expressions? Don't know.

> I'm also not quite sure what it's trying to do, is it trying to translate [...]

You should be sure, it is what it is doing :wink: 

> That also duplicates the TRANSLATED_ATTRS loop whereas it seems to me we could just... update the protocol of TRANSLATED_ATTRS to add an optional predicate?

Good point, I guess FJA did not wanted to change that `TRANSLATED_ATTRS` syntax... would you mix string and tuple with an optional predicate? Actually `TRANSLATED_ATTRS` is used by extra code, for example for view inheritance validation to avoid xpath'ing translated attribute: `TRANSLATED_ATTRS_RE = re.compile(r"@(%s)\b" % "|".join(TRANSLATED_ATTRS))`... Not sure how our conditional translation could fit in that system... simply preventing xpath on element which **might** be translatable too?

> text is the default value for `input/@type`, so should [...]

Not sure about that one. As the default type is left to browser (mis)implementation (IE's default was "button" at some point I think?), I feel like you should always specify `text` to be sure... so I would not mind the translatable part to be "opt-in" with the condition of explicitely defined `text` type. But if I change into lambda expression like suggested above, it is easy to consider omitted-types inputs too if you want :wink:

> texarea are the multi-line dual to text inputs, should we not also translate that then?

Yes, I have heard of textarea :smile: Default textarea values are their inner HTML, they do not have a `value` attribute :wink:

Reply:
> @xmo-odoo
> 
> > That doesn't seem correct: because the xpath expression is absolute [...]
> 
> Not an expert here but since `result` is a unique node, I think it is ok?

Yeah but in that case it seems like a strange method? lxml's xpath API is not really set up as a predicate system.

> I tested (again and again) and it still seems to work. If you prefer, instead of xpath expressions, I could use lambda expressions? Don't know.

I think that'd be less confusing, lxml's `xpath` API is generally designed (and used) for searching stuff, if we're checking properties of a specific node doing that in Python would be clearer.
 
> > I'm also not quite sure what it's trying to do, is it trying to translate [...]
> 
> You should be sure, it is what it is doing wink

Cool I'm not completely stupid :+1: 

> > That also duplicates the TRANSLATED_ATTRS loop whereas it seems to me we could just... update the protocol of TRANSLATED_ATTRS to add an optional predicate?
> 
> Good point, I guess FJA did not wanted to change that `TRANSLATED_ATTRS` syntax... would you mix string and tuple with an optional predicate? Actually `TRANSLATED_ATTRS` is used by extra code, for example for view inheritance validation to avoid xpath'ing translated attribute: `TRANSLATED_ATTRS_RE = re.compile(r"@(%s)\b" % "|".join(TRANSLATED_ATTRS))`... Not sure how our conditional translation could fit in that system... simply preventing xpath on element which **might** be translatable too?

Maybe just extending TRANSLATED_ATTRS to be a dict?

> 
> > text is the default value for `input/@type`, so should [...]
> 
> Not sure about that one. As the default type is left to browser (mis)implementation (IE's default was "button" at some point I think?)

I don't think `<input>` has ever been anything other than `text` by default. A bare `<input>` couldn't really be a blank button could it? It's possible that `<button>`'s default has varied I guess but that seems doubtful.

> > texarea are the multi-line dual to text inputs, should we not also translate that then?
> 
> Yes, I have heard of textarea smile Default textarea values are their inner HTML, they do not have a `value` attribute wink

Yeah what I meant is if we translate text input surely it would make sense to also translate textareas (as they're the block version), but textarea doesn't seem to be (or have been added) to the set of translated elements.

Reply:
> Maybe just extending TRANSLATED_ATTRS to be a dict?

Ok, I can change to a dict and change all related code. I'll also prevent inheritance on attributes which *might* be translatable... seems safer but maybe "migration-risky"...

> It's possible that <button>'s default has varied I guess but that seems doubtful.

Oh yes, I was confused. This is actually buttons default type which was browser dependent at some point.

> Yeah what I meant is if we translate text input surely it would make sense to also translate textareas

Yes, it works ;) Actually, this is already a translatable term in current master because it is normal text content of a block node. However if you try to translate it from the editor, it looks like that: 
![Screenshot from 2021-02-16 11-11-55](https://user-images.githubusercontent.com/10338094/108048937-f1674d00-7047-11eb-8df9-32f366f6392d.png)

... fail. This branch fixes that on the JS side.